runtime/js/fs_fake.js

```diff
@@ -24,425 +24,479 @@
-  return /\/$/.test(name) ? name : name + "/";
-};
-MlFakeDevice.prototype.lookup = function (name) {
-  if (!this.content[name] && this.lookupFun) {
-    var res = this.lookupFun(
-      caml_string_of_jsstring(this.root),
-      caml_string_of_jsstring(name),
-    );
-    if (res !== 0) {
-      this.create_dir_if_needed(name);
-      this.content[name] = new MlFakeFile(caml_bytes_of_string(res[1]));
+class MlFakeDevice {
+  constructor(root, f) {
+    this.content = {};
+    this.root = root;
+    this.lookupFun = f;
+  }
+
+  nm(name) {
+    return this.root + name;
+  }
+
+  create_dir_if_needed(name) {
+    const comp = name.split("/");
+    let res = "";
+    for (let i = 0; i < comp.length - 1; i++) {
+      res += comp[i] + "/";
+      if (this.content[res]) continue;
+      this.content[res] = Symbol("directory");
     }
   }
-};
-MlFakeDevice.prototype.exists = function (name, do_not_lookup) {
-  // The root of the device exists
-  if (name === "") return 1;
-  // Check if a directory exists
-  var name_slash = this.slash(name);
-  if (this.content[name_slash]) return 1;
-  // Check if a file exists
-  if (!do_not_lookup) this.lookup(name);
-  return this.content[name] ? 1 : 0;
-};
-MlFakeDevice.prototype.isFile = function (name) {
-  if (this.exists(name) && !this.is_dir(name)) {
-    return 1;
-  } else {
-    return 0;
+
+  slash(name) {
+    return /\/$/.test(name) ? name : name + "/";
   }
-};
-MlFakeDevice.prototype.mkdir = function (name, mode, raise_unix) {
-  if (this.exists(name))
-    caml_raise_system_error(
-      raise_unix,
-      "EEXIST",
-      "mkdir",
-      "file already exists",
-      this.nm(name),
-    );
-  var parent = /^(.*)\/[^/]+/.exec(name);
-  parent = (parent && parent[1]) || "";
-  if (!this.exists(parent))
-    caml_raise_system_error(
-      raise_unix,
-      "ENOENT",
-      "mkdir",
-      "no such file or directory",
-      this.nm(name),
-    );
-  if (!this.is_dir(parent))
-    caml_raise_system_error(
-      raise_unix,
-      "ENOTDIR",
-      "mkdir",
-      "not a directory",
-      this.nm(name),
-    );
-  this.create_dir_if_needed(this.slash(name));
-};
-MlFakeDevice.prototype.rmdir = function (name, raise_unix) {
-  var name_slash = name === "" ? "" : this.slash(name);
-  var r = new RegExp("^" + name_slash + "([^/]+)");
-  if (!this.exists(name))
-    caml_raise_system_error(
-      raise_unix,
-      "ENOENT",
-      "rmdir",
-      "no such file or directory",
-      this.nm(name),
-    );
-  if (!this.is_dir(name))
-    caml_raise_system_error(
-      raise_unix,
-      "ENOTDIR",
-      "rmdir",
-      "not a directory",
-      this.nm(name),
-    );
-  for (var n in this.content) {
-    if (n.match(r))
-      caml_raise_system_error(
-        raise_unix,
-        "ENOTEMPTY",
-        "rmdir",
-        "directory not empty",
-        this.nm(name),
+
+  lookup(name) {
+    if (!this.content[name] && this.lookupFun) {
+      const res = this.lookupFun(
```

Comment:
the `var` -> `const` rewrite doesn't seem related

Reply:
Fixed